#This code is intended to mak the final publication quality versions of the figures for this manuscript

# General Data ------------------------------------------------------------
#necessary libraries
library(phangorn)
library(ggplot2)
library(ggrepel)
library(pryr)
library(cowplot)
theme_set(theme_cowplot())
library(gt)
library(tidyverse)

#Read in the species assignment data
species <- read.csv("phylo/JTCoords.csv", header=T, stringsAsFactors = F)
species$Species <- gsub(pattern = "\\.| " , "", species$Species)
species$Species <- substr(species$Species,1,4)
species$Species <- factor(species$Species, levels=c("Aapp", "Agig", "Atec", "Ahyb", "Unkn"))

# make a color blind friendly palette
#make it from several lists so that I can call it with base plotting and ggplot
speciescolors <- c("#000000FF", 
                   "#2A788EFF",
                   "#FFCC00FF", 
                   "#808080FF", 
                   "#7AD151FF")
speciesnames <- c(expression(paste(italic("A. appalachiana"))),
                  expression(paste(italic("A. gigantea"))), 
                  expression(paste(italic("A. tecta"))), 
                  "Hybrid",
                  "Hull Road")
aruncol <- scale_color_manual(breaks = levels(species$Species),
                              values = speciescolors,
                              labels = speciesnames)

#add the colors to the species table
species$color <- setNames(speciescolors, levels(species$Species))[species$Species]

# shape hybrid samples differently
arunshape <- scale_shape_manual(breaks = levels(species$Species),
                                values = c(16, 16, 16, 4, 16),
                                labels = speciesnames)

# MDS Plots ---------------------------------------------------------------
#Read in the previously saved nuclear MDS data as generated by plink
nucmds <- read.table("phylo/nuclear/Nuclear.mds", header=T, stringsAsFactors = F)
nucmds <- merge(nucmds, species, by.x="FID", by.y="Sample")

#generate plots with and without legends
NucplotLegend <- ggplot(data = nucmds, aes(x = C1, y = C2, col=Species, shape=Species)) + 
  labs(title="Nuclear Loci \n(Ambiguous)", x="Dimension 1", y="Dimension 2") +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(hjust = 0.5),
        legend.text.align = 0) +
  geom_point(size = 3) +
  #geom_text_repel(aes(label = FID)) +
  aruncol +
  arunshape
Nucplot <- NucplotLegend + 
  theme(legend.position = "none")

# Read in the phased WXY, LFY, and plastid distance matrices from SplitsTree4
#consider whether you want to use Hamming distance (like Plink) or a phylogenetic distance
WXYmds <- read.nexus.dist(file="phylo/splitstree/WXY.dist.nex")
LFYmds <- read.nexus.dist(file="phylo/splitstree/LFY.dist.nex")
Plasmds <- read.nexus.dist(file="phylo/splitstree/Plastid_NoMissing.dist.nex")
Allmds <- read.nexus.dist(file="phylo/concatenated/TESTconatenated.dist")

#Function to process distance data into MDS
Dist2MDS <- function(dist,
                     Species=species, 
                     PlotTitle="", 
                     labels=FALSE, 
                     shapevec=arunshape, 
                     colorvec=aruncol, 
                     SizeByCount=FALSE, 
                     gridded=FALSE, 
                     facet.by=~Species,
                     jitterFactor=0) {
  #Yo, this has no error control, so don't go crazy
  fit <- cmdscale(dist, eig=T, k=2)
  vec <- cbind(rownames(fit$points),fit$points[,1:2])
  vec <- merge(vec, species, by.x="V1", by.y="Sample") #merge with species name
  vec$V2 <- as.numeric(as.character(vec$V2)) #change col types
  vec$V3 <- as.numeric(as.character(vec$V3)) #change col types
  vec$V2 <- jitter(vec$V2, factor=jitterFactor)
  vec$V3 <- jitter(vec$V3, factor=jitterFactor)
  plot <- ggplot(data=vec, aes(x = V2, y = V3, color=Species, shape=Species)) + 
    labs(title=PlotTitle, x="Dimension 1", y="Dimension 2") +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_blank(),
          axis.ticks = element_blank(),
          plot.title = element_text(hjust = 0.5),
          legend.position = "none") +
    {if(gridded)facet_wrap(facets=facet.by, ncol=1)} +
    {if(SizeByCount)geom_count()} + 
    {if(!SizeByCount)geom_point(size = 3)} +
    colorvec + 
    shapevec +
    if (labels) {
      geom_text_repel(aes(label = V1))
      }
  return(plot)
}

#Process WXY, LFY, and Plastid Data
WXYplot <- Dist2MDS(WXYmds, PlotTitle="WXY SNPs \n(Phased)", jitterFactor=300)
LFYplot <- Dist2MDS(LFYmds, PlotTitle ="LFY SNPs \n(Phased)")
PlasPlot <- Dist2MDS(Plasmds, PlotTitle = "Plastid SNPs \n(Phased)", jitterFactor = 700)
AllPlot <- Dist2MDS(Allmds, PlotTitle = "All SNPs \n(Ambiguous)", SizeByCount = F)

#Make Final Plot
legend <- get_legend(NucplotLegend + 
                       theme(legend.position = "bottom",
                             legend.justification="center",
                             legend.title = element_blank()))
right_side <- plot_grid(PlasPlot, WXYplot,nrow=2, labels=c("B", "C"))
main <- plot_grid(Nucplot, right_side, ncol=2, labels=c("A"))
plot_grid(main, legend, nrow=2, rel_heights = c(1,0.1))

ggsave2(filename = "Figure 1.pdf", height = 6,width=9)
ggsave2(filename = "Figure 1.png", height = 6, width=9)

plot_grid(LFYplot, legend, nrow=2, rel_heights = c(1,0.1))
ggsave2(filename = "Supplemental 2 LFY MDS.pdf", height=7, width=9)
ggsave2(filename = "Supplemental 2 LFY MDS.png", height=7, width=9)


# SplitsTrees -------------------------------------------------------------
# Make color-coded splitstree plots of the datasets
#These are imported from SplitsTree4
#P edulis has been removed to ease plotting
# HKY distances have been used

#Define function to clean the tip labels
TipClean <- function(network, acronym=FALSE){
  network$tip.label <- gsub(pattern="Gig$", "Gig9", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="Tec$", "Tec7", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="H_", "H-", network$tip.label)
  network$tip.label <- gsub(pattern="^igantea", "A.gigantea", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="^ecta", "A.tecta", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="^ppalachiana", "A.appalachiana", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="^A._", "A.", network$tip.label)
  if (acronym) {
    #make all hull samples "H"
    network$tip.label <- gsub(pattern="^H\\-?\\d+\\w\\_?\\d?", "H", network$tip.label, perl=T) 
    #make everything else except known samples "X"
    network$tip.label <- gsub(pattern="(^L|^Lo|^JT)\\d+\\_?\\d?", "A", network$tip.label, perl=T)
  }
  return(network)
}

#plastid data
SplitsPlas <- read.nexus.networx("phylo/splitstree/Plastid_NoMissing.splits.nex")
SplitsPlas <- TipClean(SplitsPlas)
SplitsPlasAcro <- TipClean(SplitsPlas, acronym = T)

#LFY data
SplitsLFY <- read.nexus.networx("phylo/splitstree/LFY.splits.nex")
SplitsLFY$.plot$vertices <- cbind(SplitsLFY$.plot$vertices[,"y"],
                                 -SplitsLFY$.plot$vertices[,"x"]) # Rotate 90ÂºCW
SplitsLFY <- TipClean(SplitsLFY)
SplitsLFYAcro <- TipClean(SplitsLFY, acronym = T)

#WXY data
SplitsWXY <- read.nexus.networx("phylo/splitstree/WXY.splits.nex")
SplitsWXY <- TipClean(SplitsWXY)
SplitsWXYAcro <- TipClean(SplitsWXY, acronym = T)

#Nuclear Ambiguous data
SplitsNuclear <- read.nexus.networx("phylo/splitstree/Nuclear_ambig.fasta.splits.nex")
SplitsNuclearAcro <- TipClean(SplitsNuclear, acronym=T)

#All Ambiguous Data
SplitsAll <- read.nexus.networx("phylo/splitstree/TESTconcatented.splits.nex")
SplitsAll <- TipClean(SplitsAll)
SplitsAllAcro <- TipClean(SplitsAll, acronym = T)

####Make Plots
# with help from https://www.andrewheiss.com/blog/2016/12/08/save-base-graphics-as-pseudo-objects-in-r/
#Plastid
splits.plas %<a-% {
  plot(SplitsPlasAcro, "2D",
       tip.col = species[match(SplitsPlas$tip.label, species$Sample),"color"],
       edge.width = 1.5, 
       cex = 1, 
       font = 1)
  mtext("Plastid", side=3, cex=1.7, line=-1)
}

#Plot LFY 
splits.LFY %<a-% {
  plot(SplitsLFYAcro, "2D",
       tip.col = species[match(SplitsLFY$tip.label, species$Sample),"color"],
       edge.width = .7, 
       cex = 1, 
       font = 3)
  #there is a gap btw Tec and Gig that separates each Hull Haplotype
  mtext("LFY", side=3, cex=1.7, line=-1)
  arrows(0.00625,-0.003125,-0.001,-0.001, lwd=3, col="red", angle=15) #for upright plot
  #arrows(-0.0045,-0.0035,-0.001,0.001, lwd=3, col="red", angle=15) #for flat plot
}

#Plot WXY
splits.WXY %<a-% {
  plot(SplitsWXYAcro, "2D",
     tip.col = species[match(SplitsWXY$tip.label, species$Sample),"color"],
     edge.width = 1, 
     cex = 1, 
     font = 3)
  mtext("WXY", side=3, cex=1.7, line=-1)
}

#Plot Nuclear
splits.Nuclear %<a-% {
  plot(SplitsNuclearAcro, "2D",
     tip.col = species[match(SplitsNuclear$tip.label, species$Sample),"color"],
     edge.width = 1.5, 
     cex = 1, 
     font = 3)
  mtext("Nuclear", side=3, cex=1.7, line=-1)
}

#Plot All
splits.All %<a-% {
  plot(SplitsAllAcro, "2D",
     tip.col = species[match(SplitsAll$tip.label, species$Sample),"color"],
     edge.width = 1.5, 
     cex = 1, 
     font = 1)
  mtext("All Loci", side=3, cex=1.7, line=-1)
}

#Figure 2
Figure2 %<a-% {
  par(mar=c(1,0,1,0))
  split.screen(c(1,2))
  par(oma=c(0,0,2,0))
  split.screen(c(3,1), screen=2)
  #split.screen(c(1,2), screen=4)
  screen(3) #maybe switch to screen 1?
  splits.All
  screen(1) #aybe switch to screen 3
  splits.LFY
  legend("bottomright",legend=speciesnames,col=speciescolors, pch=c(rep("A",4),"H"), bty="n", cex=1.3)
  screen(4)
  splits.WXY
  screen(5)
  splits.plas
  mtext("SplitsTree", side=3, outer=T, line=, cex=3)
  close.screen(all=T)
}

#Save Figure 2
cairo_pdf(file="Figure 2 Alt.pdf", height=11, width=15)
  Figure2
dev.off()

#consider an alternative where SplitsAll is rotated, SPlitsLFY isnt and the two are swapped in the figure to show the division in LFY better



# Morphological Table -----------------------------------------------------
morpho <- read.csv2("Morpho.csv", header=T, sep=",")
#system("PATH=/bigdata/littlab/arajewski/Datura/software/phantomjs-2.1.1-linux-x86_64/bin:$PATH")
morpho %>%
  gt(
    rownames_to_stub = T
  ) %>%
  tab_header(
    title = "Morphological Comparison") %>%
  cols_align(
    align="right"
    ) %>%
  tab_style(
    style = list(cell_text(style="italic")),
    locations = cells_stub(rows=c(1:3))) %>%
  cols_label(
    Aerenchyma = md("**Aerenchyma</br>in Rhizomes**"),
    Sulcus = md("**Grooved</br>Internodes**"),
    Internodes = md("**Internodes</br>at Branch</br>Complement**"),
    Pubescence = md("**Abaxial Leaf</br>Pubescence**")) %>%
  tab_footnote(
    footnote="Triplett et al, 2006",
    locations = cells_stub(
      rows = c(1:3))) %>%
  tab_footnote(
    footnote="This study",
    locations = cells_stub(
      rows = c(4:5))) %>%
  gtsave(filename = "Morpho_table.pdf")

# Supplemental Phylogenetic Trees -----------------------------------------
#read in the ambiguous, unphased, concatenated RAxML tree
#RAxMLPhylo <- read.newick("concatenated/GTRGI/TESTconcatenated.phy.partitioned.raxml.support")
#or read in tree collapsed by <10% bootstrap in Tree Graph
RAxMLPhylo <- read.newick("phylo/concatenated/GTRGI/TESTconcatenated.phy.partitioned.raxml.collapsed.nwk")
#clean the labels to match
RAxMLPhylo$tip.label <- gsub(pattern = "\\_" , " ", RAxMLPhylo$tip.label)
RAxMLPhylo$tip.label <- gsub(pattern = "Tec" , "Tec7", RAxMLPhylo$tip.label)
RAxMLPhylo$tip.label <- gsub(pattern = "H 2C" , "H-2C", RAxMLPhylo$tip.label)
RAxMLPhylo$tip.label <- gsub(pattern = "H 2D" , "H-2D", RAxMLPhylo$tip.label)
RAxMLPhylo$tip.label <- gsub(pattern = "Gig" , "Gig9", RAxMLPhylo$tip.label)
RAxMLPhylo$node.label <- gsub(pattern = "\\.\\d+$", "", RAxMLPhylo$node.label)

#snippet to sort the tips' colors vector (repeated below)
species[match(RAxMLPhylo$tip.label, species$Sample),5]
#make a vector of shape numbers. open cirle = BS 1-25, closed circle = BS >25
RAxMLShapes <- RAxMLPhylo$node.label
RAxMLShapes[RAxMLPhylo$node.label<=25] <- 1
RAxMLShapes[RAxMLPhylo$node.label>25] <- 16

#and plot it
#This is currently (21 Jan 20) Supplemental Figure 1
pdf(file="phylo/concatenated/GTRGI/TESTconcatenated.phy.partitioned.raxml.support.pdf",
    width=10,
    height=22)
  plot(RAxMLPhylo, 
       show.node.label = F,
       tip.color=species[match(RAxMLPhylo$tip.label, species$Sample),5],
       font=1)
  nodelabels(pch=as.numeric(RAxMLShapes))
  legend("bottomright",
         legend=c(speciesnames, "10-25% Bootstrap", ">25% Bootstrap", "", ""),
         col=c(speciescolors,"black", "black"), 
         pch=c(NA, NA, NA, NA, NA, 1, 16),
         bty="n",
         cex=0.9,
         ncol=2,
         text.col=c(speciescolors,"black", "black"))
dev.off()
