#This code is intended to mak the final publication quality versions of the figures for this manuscript

# General Data ------------------------------------------------------------
#necessary libraries
library(phangorn)
library(ggplot2)
library(ggrepel)
library(cowplot)
theme_set(theme_cowplot())

#Read in the species assignment data
species <- read.csv("phylo/JTCoords.csv", header=T, stringsAsFactors = F)
species$Species <- gsub(pattern = "\\.| " , "", species$Species)
species$Species <- substr(species$Species,1,4)
species$Species <- factor(species$Species, levels=c("Aapp", "Agig", "Atec", "Ahyb", "Unkn"))

# make a color blind friendly palette
#make it from several lists so that I can call it with base plotting and ggplot
speciescolors <- c("#000000FF", 
                   "#2A788EFF",
                   "#FFCC00FF", 
                   "#808080FF", 
                   "#7AD151FF")
speciesnames <- c(expression(paste(italic("A. appalachiana"))),
                  expression(paste(italic("A. gigantea"))), 
                  expression(paste(italic("A. tecta"))), 
                  "Hybrid",
                  "Hull Road")
aruncol <- scale_color_manual(breaks = levels(species$Species),
                              values = speciescolors,
                              labels = speciesnames)

#add the colors to the species table
species$color <- setNames(speciescolors, levels(species$Species))[species$Species]

# shape hybrid sampels differently
arunshape <- scale_shape_manual(breaks = levels(species$Species),
                                values = c(16, 16, 16, 4, 16),
                                labels = speciesnames)


# MDS Plots ---------------------------------------------------------------
#Read in the previously saved nuclear MDS data as generated by plink
nucmds <- read.table("phylo/nuclear/Nuclear.mds", header=T, stringsAsFactors = F)
nucmds <- merge(nucmds, species, by.x="FID", by.y="Sample")

NucplotLegend <- ggplot(data = nucmds, aes(x = C1, y = C2, col=Species, shape=Species)) + 
  labs(title="Nuclear Loci \n(Ambiguous)", x="Dimension 1", y="Dimension 2") +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(hjust = 0.5),
        legend.text.align = 0) +
  geom_point(size = 3) +
  #geom_text_repel(aes(label = FID)) +
  aruncol +
  arunshape

Nucplot <- NucplotLegend + 
  theme(legend.position = "none")

# Read in the phased WXY, LFY, and plastid distance matrices from SplitsTree4
#consider whether you want to use Hamming distance (like Plink) or a phylogenetic distance
WXYmds <- read.nexus.dist(file="phylo/WXY/WXY.dist.nex")
LFYmds <- read.nexus.dist(file="phylo/LFY/LFY.dist.nex")
Plasmds <- read.nexus.dist(file="phylo/plastid/Plastid_NoMissing.dist.nex")

#Process WXY
WXYfit <- cmdscale(WXYmds, eig=T, k=2) #fit mds
WXYvec <- cbind(rownames(WXYfit$points), WXYfit$points[,1:2])
WXYvec <- merge(WXYvec, species, by.x="V1", by.y="Sample") #merge with species names
WXYvec$V2 <- as.numeric(as.character(WXYvec$V2)) #change col types
WXYvec$V3 <- as.numeric(as.character(WXYvec$V3)) #change col types
WXYplot <- ggplot(data =WXYvec, aes(x = V2, y = V3, color=Species, shape=Species)) + 
labs(title="WXY SNPs \n(Phased)", x="Dimension 1", y="Dimension 2") +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(hjust = 0.5),
        legend.position = "none") +
  geom_point(size = 3) +
  #geom_text_repel(data=subset(WXYvec, Species == "Unkn"), aes(label = V1)) +
  aruncol + 
  arunshape

#Process LFY
LFYfit <- cmdscale(LFYmds, eig=T, k=2) #fit mds
LFYvec <- cbind(rownames(LFYfit$points), LFYfit$points[,1:2])
LFYvec <- merge(LFYvec, species, by.x="V1", by.y="Sample") #merge with species names
LFYvec$V2 <- as.numeric(as.character(LFYvec$V2)) #change col types
LFYvec$V3 <- as.numeric(as.character(LFYvec$V3)) #change col types
LFYplot <- ggplot(data = LFYvec, aes(x = V2, y = V3, color=Species, shape=Species)) + 
  labs(title="LFY SNPs \n(Phased)", x="Dimension 1", y="Dimension 2") +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(hjust = 0.5),
        legend.position = "none") +
  geom_point(size = 3) +
  #geom_text_repel(aes(label = V1)) +
  aruncol

#Process Plastid
Plasfit <- cmdscale(Plasmds, eig=T, k=2) #fit mds
Plasvec <- cbind(rownames(Plasfit$points), Plasfit$points[,1:2])
Plasvec <- merge(Plasvec, species, by.x="V1", by.y="Sample") #merge with species names
Plasvec$V2 <- as.numeric(as.character(Plasvec$V2)) #change col types
Plasvec$V3 <- as.numeric(as.character(Plasvec$V3)) #change col types
PlasPlot <- ggplot(data = Plasvec, aes(x = V2, y = V3, color=Species, shape=Species)) + 
  labs(title="Plastid SNPs \n(Phased)", x="Dimension 1", y="Dimension 2") +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(hjust = 0.5),
        legend.position = "none") +
  geom_point(size = 3) +
  #geom_text_repel(aes(label = V1)) +
  aruncol +
  arunshape

#Make Final Plot
legend <- get_legend(NucplotLegend + 
                       theme(legend.position = "bottom",
                             legend.justification="center",
                             legend.title = element_blank()))
right_side <- plot_grid(PlasPlot, WXYplot,nrow=2, labels=c("B", "C"))
main <- plot_grid(Nucplot, right_side, ncol=2, labels=c("A"))
plot_grid(main, legend, nrow=2, rel_heights = c(1,0.1))

ggsave2(filename = "Figure 1.pdf", height = 6,width=9)


# SplitsTrees -------------------------------------------------------------
# Make color-coded splitstree plots of the datasets
#These are imported from SplitsTree4
#P edulis has been removed to ease plotting
# HKY distances have been used

#Define function to clean the tip labels
TipClean <- function(network, acronym=FALSE){
  network$tip.label <- gsub(pattern="$Gig", "Gig9", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="$Tec", "Tec7", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="H_", "H-", network$tip.label)
  network$tip.label <- gsub(pattern="^igantea", "A.gigantea", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="^ecta", "A.tecta", network$tip.label, perl=T)
  network$tip.label <- gsub(pattern="^ppalachiana", "A.appalachiana", network$tip.label, perl=T)
  if (acronym) {
    #make all hull samples "H"
    network$tip.label <- gsub(pattern="^H\\-?\\d+\\w\\_?\\d?", "H", network$tip.label, perl=T) 
    #make everything else except known samples "X"
    network$tip.label <- gsub(pattern="(^L|^Lo|^JT)\\d+\\_?\\d?", "X", network$tip.label, perl=T)
  }
  return(network)
}

#plastid data
SplitsPlas <- read.nexus.networx("phylo/splitstree/Plastid_NoMissing.splits.nex")
SplitsPlas <- TipClean(SplitsPlas)
SplitsPlasAcro <- TipClean(SplitsPlas, acronym = T)

#LFY data
SplitsLFY <- read.nexus.networx("phylo/splitstree/LFY.splits.nex")
SplitsLFY$.plot$vertices <- cbind(-SplitsLFY$.plot$vertices[,"y"],
                                  SplitsLFY$.plot$vertices[,"x"]) # Rotate 90ÂºCCW
SplitsLFY <- TipClean(SplitsLFY)
SplitsLFYAcro <- TipClean(SplitsLFY, acronym = T)

#WXY data
SplitsWXY <- read.nexus.networx("phylo/splitstree/WXY.splits.nex")
SplitsWXY <- TipClean(SplitsWXY)
SplitsWXYAcro <- TipClean(SplitsWXY, acronym = T)

#Nuclear Ambiguous data
SplitsNuclear <- read.nexus.networx("phylo/splitstree/Nuclear_ambig.fasta.splits.nex")

#All Ambiguous Data
SplitsAll <- read.nexus.networx("phylo/splitstree/TESTconcatented.splits.nex")
SplitsAll <- TipClean(SplitsAll)
SplitsAllAcro <- TipClean(SplitsAll, acronym = T)

#Plot Plastid
par(mar = c(0,0,1,0))
plot(SplitsPlasDots, "2D",
     tip.col = species[match(SplitsPlas$tip.label, species$Sample),"color"],
     edge.width = 1.5, 
     cex = 1, 
     font = 1)
title(main = "Plastid SNP SplitsTree")
legend("topleft",legend=speciesnames,col=speciescolors, pch=16, bty="n")
dev.off()

#Plot LFY 
par(mar = c(0,0,1,0))
plot(SplitsLFYRotate, "2D",
     tip.col = species[match(SplitsLFY$tip.label, species$Sample),"color"],
     edge.width = 1.5, 
     cex = 1, 
     font = 3)
#there is a gap btw Tec and Gig that separates each Hull Haplotype
title(main = "LFY SNP SplitsTree")
legend("topleft",legend=speciesnames,col=speciescolors, pch=16, bty="n")
dev.off()

#Plot WXY
par(mar = c(0,0,1,0))
plot(SplitsWXY, "2D",
     tip.col = species[match(SplitsWXY$tip.label, species$Sample),"color"],
     edge.width = 1.5, 
     cex = 1, 
     font = 3)
title(main = "WXY SNP SplitsTree")
legend("topleft",legend=speciesnames,col=speciescolors, pch=16, bty="n")
dev.off()

#Plot Nuclear
par(mar = c(0,0,1,0))
plot(SplitsNuclear, "2D",
     tip.col = species[match(SplitsNuclear$tip.label, species$Sample),"color"],
     tip.label = "",
     edge.width = 1.5, 
     cex = 1, 
     font = 3)
title(main = "Nuclear SNP SplitsTree")
legend("topleft",legend=speciesnames,col=speciescolors, pch=16, bty="n")
dev.off()

#Plot All
par(mar = c(0,0,1,0))
plot(SplitsAll, "2D",
     tip.col = species[match(SplitsAll$tip.label, species$Sample),"color"],
     edge.width = 1.5, 
     cex = 1, 
     font = 3)
title(main = "All SNP SplitsTree")
legend("topleft",legend=speciesnames,col=speciescolors, pch=16, bty="n")
dev.off()


#make the figure?
layout(matrix(c(1,1,2,1,1,2,1,1,3,1,1,3,4,4,4), nrow=5, ncol=3, byrow = T))
par(oma=c(0,1,3,0))
#insert the three plots you want, then:
mtext("SplitsTree Networks", outer=T, cex=2)
plot.new()
legend("center",legend=speciesnames,col=speciescolors, pch=16, bty="n", horiz=T, cex=1.1, x.intersp=0.5)

